proposals/0024-repair-and-restart.md

```diff
@@ -0,0 +1,175 @@
+---
+simd: '0024'
+title: Automatially repair and start for a cluster restart
+authors:
+  - Wen Xu (Solana Labs)
+category: Standard
+type: Core
+status: Draft
+created: 2023-04-07
+feature: (fill in with feature tracking issues once accepted)
+---
+
+## Summary
+
+Improve the current [cluster restart]
+(https://docs.solana.com/running-validator/restart-cluster)
+procedure such that validators can automatically figure out confirmed slots
+and then proceed to restart if everything looks fine.
+
+## Motivation
+
+Currently during a cluster restart, validator operators need to decide latest
+optimistically confirmed slot, then restart the validators with new commandline
+arguments.
+
+The current process involves a lot of human intervention, if people make a
+mistake in deciding the highest optimistically confirmed slot, it could mean
+rollback of user transactions after those transactions have been confirmed,
+which is not acceptable.
+
+We aim to automate the finding of highest optimistically confirmed slot and
+block data distribution, so that we can lower the possibility of human mistakes
+in the cluster restart process.
+
+## Alternatives Considered
+
+See [Handling Common Solana Outages]
+(https://docs.google.com/document/d/1RkNAyz-5aKvv5FF44b8SoKifChKB705y5SdcEoqMPIc)
+for details.
+
+There are many proposals about automatically detecting that the cluster is
+in an outage so validators should enter a recovery process automatically.
+
+While getting human out of the loop greatly improves recovery speed,
+automaticlly restarting the whole cluster still seems risky. Because if
+the recovery process itself doesn't work, it might be some time before
+we can get human's attention. And it doesn't solve the cases where new binary
+is needed. So for now we still plan to have human in the loop.
+
+## New Terminology
+
+* SilentRepairMode - when validators restart in this new mode, they will
+talk with each other to find the highest optimistically confirmed slot and
+repair all necessary blocks. To improve speed and guarantee simplicity,
+Turbine, vote, and new block generation are paused in this mode.
```

Comment:
```suggestion
* SilentRepairMode - when validators are started in this new mode, they will
gossip with each other to identity the highest optimistically confirmed slot and
repair all its ancestor blocks. To improve speed and guarantee simplicity,
Turbine, voting, and new block generation are paused in this mode.
```

I don't love the name "SilentRepairMode" yet, but I think we can refine that bit a little later

Reply:
How about "RepairBeforeRestartMode"?

Reply:
I'm not sure we even need to name it tbh. For example the current `solana-validator --wait-for-supermajority` argument expects a `SLOT`. In the eventual Labs implementation of this SIMD, if a `SLOT` is not provided then we could enter this automated mode of identifying and repairing to the highest oc slot.